File: isa.cc

#include <cassert>
#include <iostream>

#include "cpu.h"

namespace gvm {

Word Nop() {
  return Word(ISA::NOP);
}

Word Halt() {
  return Word(ISA::HALT);
}

Word MovRR(uint32_t dest, uint32_t src) {
  return AddRR(dest, src, 0);
}

Word MovRI(uint32_t dest, uint32_t value) {
  return AddRI(dest, 28, value);
}

Word LoadRI(uint32_t dest, uint32_t memaddr) {
  assert(dest < kRegCount);
  assert((memaddr & 0x1FFFFF) == memaddr);
  assert(memaddr % kWordSize == 0);  // kWordSize aligned memory.
  return Word(ISA::LOAD_RI | dest << 6 | memaddr << 11);
}

Word LoadRR(uint32_t dest, uint32_t src) {
  return LoadIX(dest, src, 0);
}

Word LoadIX(uint32_t dest, uint32_t src, uint32_t offset) {
  assert(dest < kRegCount);
  assert(src < kRegCount);
  offset = offset & 0xFFFF;
  return Word(ISA::LOAD_IX | dest << 6 | src << 11  | offset << 16);
}

Word LoadIXR(uint32_t dest, uint32_t src, uint32_t offset) {
  assert(dest < kRegCount);
  assert(src < kRegCount);
  assert(offset < kRegCount);
  return Word(ISA::LOAD_IXR | dest << 6 | src << 11  | offset << 16);
}

Word LoadPI(uint32_t dest, uint32_t src, uint32_t offset) {
  assert(dest < kRegCount);
  assert(src < kRegCount);
  offset = offset & 0xFFFF;
  return Word(ISA::LOAD_PI | dest << 6 | src << 11  | offset << 16);
}

Word LoadIP(uint32_t dest, uint32_t src, uint32_t offset) {
  assert(dest < kRegCount);
  assert(src < kRegCount);
  offset = offset & 0xFFFF;
  return Word(ISA::LOAD_IP | dest << 6 | src << 11  | offset << 16);
}

Word StorRI(uint32_t memaddr, uint32_t src) {
  assert((memaddr & 0x1FFFFF) == memaddr);
  assert(memaddr % kWordSize == 0);  // kWordSize aligned memory.
  assert(src < kRegCount);
  return Word(ISA::STOR_RI | src << 6 | memaddr << 11);
}

Word StorRR(uint32_t dest, uint32_t src) {
  return StorIX(dest, src, 0);
}

Word StorIX(uint32_t dest, uint32_t src, uint32_t offset) {
  assert(dest < kRegCount);
  assert(src < kRegCount);
  offset = offset & 0xFFFF;
  return Word(ISA::STOR_IX | dest << 6 | src << 11  | offset << 16);
}

Word StorPI(uint32_t dest, uint32_t src, uint32_t offset) {
  assert(dest < kRegCount);
  assert(src < kRegCount);
  offset = offset & 0xFFFF;
  return Word(ISA::STOR_PI | dest << 6 | src << 11  | offset << 16);
}

Word StorIP(uint32_t dest, uint32_t src, uint32_t offset) {
  assert(dest < kRegCount);
  assert(src < kRegCount);
  offset = offset & 0xFFFF;
  return Word(ISA::STOR_IP | dest << 6 | src << 11  | offset << 16);
}

Word AddRR(uint32_t dest, uint32_t op1, uint32_t op2) {
  assert(dest < kRegCount);
  assert(op1 < kRegCount);
  assert(op2 < kRegCount);
  return Word(ISA::ADD_RR | dest << 6 | op1 << 11 | op2 << 16);
}

Word AddRI(uint32_t dest, uint32_t op1, uint32_t value) {
  assert(dest < kRegCount);
  assert(op1 < kRegCount);
  value = value & 0xFFFF;
  return Word(ISA::ADD_RI | dest << 6 | op1 << 11 | value << 16);
}

Word SubRR(uint32_t dest, uint32_t op1, uint32_t op2) {
  assert(dest < kRegCount);
  assert(op1 < kRegCount);
  assert(op2 < kRegCount);
  return Word(ISA::SUB_RR | dest << 6 | op1 << 11 | op2 << 16);
}

Word SubRI(uint32_t dest, uint32_t op1, uint32_t value) {
  assert(dest < kRegCount);
  assert(op1 < kRegCount);
  value = value & 0xFFFF;
  return Word(ISA::SUB_RI | dest << 6 | op1 << 11 | value << 16);
}

Word Jmp(uint32_t memaddr) {
  memaddr = memaddr & 0x3FFFFFF;
  return Word(ISA::JMP | memaddr << 6);
}

Word Jne(uint32_t reg, uint32_t memaddr) {
  assert(reg < kRegCount);
  memaddr = memaddr & 0x1FFFFF;
  return Word(ISA::JNE | reg << 6 | memaddr << 11);
}

Word Jeq(uint32_t reg, uint32_t memaddr) {
  assert(reg < kRegCount);
  memaddr = memaddr & 0x1FFFFF;
  return Word(ISA::JEQ | reg << 6 | memaddr << 11);
}

Word Jgt(uint32_t reg, uint32_t memaddr) {
  assert(reg < kRegCount);
  memaddr = memaddr & 0x1FFFFF;
  return Word(ISA::JGT | reg << 6 | memaddr << 11);
}

Word Jge(uint32_t reg, uint32_t memaddr) {
  assert(reg < kRegCount);
  memaddr = memaddr & 0x1FFFFF;
  return Word(ISA::JGE | reg << 6 | memaddr << 11);
}

Word Jlt(uint32_t reg, uint32_t memaddr) {
  assert(reg < kRegCount);
  memaddr = memaddr & 0x1FFFFF;
  return Word(ISA::JLT | reg << 6 | memaddr << 11);
}

Word Jle(uint32_t reg, uint32_t memaddr) {
  assert(reg < kRegCount);
  memaddr = memaddr & 0x1FFFFF;
  return Word(ISA::JLE | reg << 6 | memaddr << 11);
}

Word CallI(uint32_t memaddr) {
  memaddr = memaddr & 0x3FFFFFF;
  return Word(ISA::CALLI | memaddr << 6);
}

Word CallR(uint32_t dest) {
  assert(dest < kRegCount);
  return Word(ISA::CALLR | dest << 6);
}

Word Ret() {
  return Word(ISA::RET);
}

Word AndRR(uint32_t dest, uint32_t op1, uint32_t op2) {
  assert(dest < kRegCount);
  assert(op1 < kRegCount);
  assert(op2 < kRegCount);
  return Word(ISA::AND_RR | dest << 6 | op1 << 11 | op2 << 16);
}

Word AndRI(uint32_t dest, uint32_t op1, uint32_t value) {
  assert(dest < kRegCount);
  assert(op1 < kRegCount);
  value = value & 0xFFFF;
  return Word(ISA::AND_RI | dest << 6 | op1 << 11 | value << 16);
}

Word OrrRR(uint32_t dest, uint32_t op1, uint32_t op2) {
  assert(dest < kRegCount);
  assert(op1 < kRegCount);
  assert(op2 < kRegCount);
  return Word(ISA::ORR_RR | dest << 6 | op1 << 11 | op2 << 16);
}

Word OrrRI(uint32_t dest, uint32_t op1, uint32_t value) {
  assert(dest < kRegCount);
  assert(op1 < kRegCount);
  value = value & 0xFFFF;
  return Word(ISA::ORR_RI | dest << 6 | op1 << 11 | value << 16);
}

Word XorRR(uint32_t dest, uint32_t op1, uint32_t op2) {
  assert(dest < kRegCount);
  assert(op1 < kRegCount);
  assert(op2 < kRegCount);
  return Word(ISA::XOR_RR | dest << 6 | op1 << 11 | op2 << 16);
}

Word XorRI(uint32_t dest, uint32_t op1, uint32_t value) {
  assert(dest < kRegCount);
  assert(op1 < kRegCount);
  value = value & 0xFFFF;
  return Word(ISA::XOR_RI | dest << 6 | op1 << 11 | value << 16);
}

Word LslRR(uint32_t dest, uint32_t src, uint32_t offset) {
  assert(dest < kRegCount);
  assert(src < kRegCount);
  assert(offset < kRegCount);
  return Word(ISA::LSL_RR | dest << 6 | src << 11 | offset  << 16);
}

Word LslRI(uint32_t dest, uint32_t src, uint32_t value) {
  assert(dest < kRegCount);
  assert(src < kRegCount);
  value = value & 0xFFFF;
  return Word(ISA::LSL_RI | dest << 6 | src << 11 | value  << 16);
}

Word LsrRR(uint32_t dest, uint32_t src, uint32_t offset) {
  assert(dest < kRegCount);
  assert(src < kRegCount);
  assert(offset < kRegCount);
  return Word(ISA::LSR_RR | dest << 6 | src << 11 | offset  << 16);
}

Word LsrRI(uint32_t dest, uint32_t src, uint32_t value) {
  assert(dest < kRegCount);
  assert(src < kRegCount);
  value = value & 0xFFFF;
  return Word(ISA::LSR_RI | dest << 6 | src << 11 | value  << 16);
}

Word AsrRR(uint32_t dest, uint32_t src, uint32_t offset) {
  assert(dest < kRegCount);
  assert(src < kRegCount);
  assert(offset < kRegCount);
  return Word(ISA::ASR_RR | dest << 6 | src << 11 | offset  << 16);
}

Word AsrRI(uint32_t dest, uint32_t src, uint32_t value) {
  assert(dest < kRegCount);
  assert(src < kRegCount);
  value = value & 0xFFFF;
  return Word(ISA::ASR_RI | dest << 6 | src << 11 | value  << 16);
}

Word MulRR(uint32_t dest, uint32_t op1, uint32_t op2) {
  assert(dest < kRegCount);
  assert(op1 < kRegCount);
  assert(op2 < kRegCount);
  return Word(ISA::MUL_RR | dest << 6 | op1 << 11 | op2  << 16);
}

Word MulRI(uint32_t dest, uint32_t op1, uint32_t value) {
  assert(dest < kRegCount);
  assert(op1 < kRegCount);
  value = value & 0xFFFF;
  return Word(ISA::MUL_RI | dest << 6 | op1 << 11 | value  << 16);
}

Word DivRR(uint32_t dest, uint32_t op1, uint32_t op2) {
  assert(dest < kRegCount);
  assert(op1 < kRegCount);
  assert(op2 < kRegCount);
  return Word(ISA::DIV_RR | dest << 6 | op1 << 11 | op2  << 16);
}

Word DivRI(uint32_t dest, uint32_t op1, uint32_t value) {
  assert(dest < kRegCount);
  assert(op1 < kRegCount);
  value = value & 0xFFFF;
  return Word(ISA::DIV_RI | dest << 6 | op1 << 11 | value  << 16);
}

Word MullRR(uint32_t destH, uint32_t destL, uint32_t op1, uint32_t op2) {
  assert(destH < kRegCount);
  assert(destL < kRegCount);
  assert(op1 < kRegCount);
  assert(op2 < kRegCount);
  return Word(ISA::MULL_RR | destH << 6 | destL << 11 | op1 << 16 | op2 << 21);
}

Word Wfi() {
  return Word(ISA::WFI);
}

}  // namespace gvm

File: gsm/code/isa.go
/*
 * Copyright (C) 2019  Igor Cananea <icc@avalonbits.com>
 * Author: Igor Cananea <icc@avalonbits.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package code

import (
	"github.com/avalonbits/gsm/parser"
)

func Nop() parser.Word {
	return parser.Word(parser.Nop)
}

func Ret() parser.Word {
	return parser.Word(parser.Ret)
}

func Halt() parser.Word {
	return parser.Word(parser.Halt)
}

func MovRR(dest, src uint32) parser.Word {
	return AddRI(dest, src, 0)
}

func MovRI(dest, value uint32) parser.Word {
	return AddRI(dest, 31, value)
}

func LoadRR(dest, src uint32) parser.Word {
	return LoadIX(dest, src, 0)
}

func LoadPC(dest, value uint32) parser.Word {
	value = value & 0x1FFFFF
	return parser.Word(parser.Load_pc) | parser.Word(dest)<<6 | parser.Word(value)<<11
}

func LoadRI(dest, value uint32) parser.Word {
	value = value & 0x1FFFFF
	return parser.Word(parser.Load_ri) | parser.Word(dest)<<6 | parser.Word(value)<<11
}

func LoadIX(dest, src, offset uint32) parser.Word {
	offset = offset & 0xFFFF
	return parser.Word(parser.Load_ix) | parser.Word(dest)<<6 | parser.Word(src)<<11 | parser.Word(offset)<<16
}

func LoadIXR(dest, src, offset uint32) parser.Word {
	return parser.Word(parser.Load_ixr) | parser.Word(dest)<<6 | parser.Word(src)<<11 |
		parser.Word(offset)<<16
}

func LoadIP(dest, src, offset uint32) parser.Word {
	offset = offset & 0xFFFF
	return parser.Word(parser.Load_ip) | parser.Word(dest)<<6 | parser.Word(src)<<11 | parser.Word(offset)<<16
}

func LoadPairIP(dest1, dest2, src, offset uint32) parser.Word {
	offset = offset & 0x7FF
	return parser.Word(parser.Ldp_ip) | parser.Word(dest1)<<6 | parser.Word(dest2)<<11 |
		parser.Word(src)<<16 | parser.Word(offset)<<21
}

func LoadPI(dest, src, offset uint32) parser.Word {
	offset = offset & 0xFFFF
	return parser.Word(parser.Load_pi) | parser.Word(dest)<<6 | parser.Word(src)<<11 | parser.Word(offset)<<16
}

func LoadPairPI(dest1, dest2, src, offset uint32) parser.Word {
	offset = offset & 0x7FF
	return parser.Word(parser.Ldp_pi) | parser.Word(dest1)<<6 | parser.Word(dest2)<<11 |
		parser.Word(src)<<16 | parser.Word(offset)<<21
}

func StorRR(dest, src uint32) parser.Word {
	return StorIX(dest, src, 0)
}

func StorRI(dest, src uint32) parser.Word {
	dest = dest & 0x1FFFFF
	return parser.Word(parser.Stor_ri) | parser.Word(src)<<6 | parser.Word(dest)<<11
}

func StorIX(dest, src, offset uint32) parser.Word {
	offset = offset & 0xFFFF
	return parser.Word(parser.Stor_ix) | parser.Word(dest)<<6 | parser.Word(src)<<11 | parser.Word(offset)<<16
}

func StorPC(dest, src uint32) parser.Word {
	dest = dest & 0x1FFFFF
	return parser.Word(parser.Stor_pc) | parser.Word(src)<<6 | parser.Word(dest)<<11
}

func StorPI(dest, src, offset uint32) parser.Word {
	offset = offset & 0xFFFF
	return parser.Word(parser.Stor_pi) | parser.Word(dest)<<6 | parser.Word(src)<<11 | parser.Word(offset)<<16
}

func StorIP(dest, src, offset uint32) parser.Word {
	offset = offset & 0xFFFF
	return parser.Word(parser.Stor_ip) | parser.Word(dest)<<6 | parser.Word(src)<<11 | parser.Word(offset)<<16
}

func StorPairPI(dest, src1, src2, offset uint32) parser.Word {
	offset = offset & 0x7FF
	return parser.Word(parser.Stp_pi) | parser.Word(dest)<<6 | parser.Word(src1)<<11 |
		parser.Word(src2)<<16 | parser.Word(offset)<<21
}

func StorPairIP(dest, src1, src2, offset uint32) parser.Word {
	offset = offset & 0x7FF
	return parser.Word(parser.Stp_ip) | parser.Word(dest)<<6 | parser.Word(src1)<<11 | parser.Word(src2)<<16 | parser.Word(offset)<<21
}

func AddRR(dest, op1, op2 uint32) parser.Word {
	return parser.Word(parser.Add_rr) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func AddRI(dest, op1, op2 uint32) parser.Word {
	op2 = op2 & 0xFFFF
	return parser.Word(parser.Add_ri) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func SubRR(dest, op1, op2 uint32) parser.Word {
	return parser.Word(parser.Sub_rr) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func SubRI(dest, op1, op2 uint32) parser.Word {
	op2 = op2 & 0xFFFF
	return parser.Word(parser.Sub_ri) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func Jmp(addr uint32) parser.Word {
	addr = addr & 0x3FFFFFF
	return parser.Word(parser.Jmp) | parser.Word(addr)<<6
}

func jumpc(inst parser.OpCode, reg, addr uint32) parser.Word {
	addr = addr & 0x1FFFFF
	return parser.Word(inst) | parser.Word(reg)<<6 | parser.Word(addr)<<11
}

func Jne(reg, addr uint32) parser.Word {
	return jumpc(parser.Jne, reg, addr)
}

func Jeq(reg, addr uint32) parser.Word {
	return jumpc(parser.Jeq, reg, addr)
}

func Jgt(reg, addr uint32) parser.Word {
	return jumpc(parser.Jgt, reg, addr)
}

func Jge(reg, addr uint32) parser.Word {
	return jumpc(parser.Jge, reg, addr)
}

func Jlt(reg, addr uint32) parser.Word {
	return jumpc(parser.Jlt, reg, addr)
}

func Jle(reg, addr uint32) parser.Word {
	return jumpc(parser.Jle, reg, addr)
}

func CallI(addr uint32) parser.Word {
	addr = addr & 0x3FFFFFF
	return parser.Word(parser.Call_i) | parser.Word(addr)<<6
}

func CallR(dest uint32) parser.Word {
	return parser.Word(parser.Call_r) | parser.Word(dest)<<6
}

func AndRR(dest, op1, op2 uint32) parser.Word {
	return parser.Word(parser.And_rr) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func OrrRR(dest, op1, op2 uint32) parser.Word {
	return parser.Word(parser.Orr_rr) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func XorRR(dest, op1, op2 uint32) parser.Word {
	return parser.Word(parser.Xor_rr) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func LslRR(dest, op1, op2 uint32) parser.Word {
	return parser.Word(parser.Lsl_rr) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func LsrRR(dest, op1, op2 uint32) parser.Word {
	return parser.Word(parser.Lsr_rr) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func AsrRR(dest, op1, op2 uint32) parser.Word {
	return parser.Word(parser.Asr_rr) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func MulRR(dest, op1, op2 uint32) parser.Word {
	return parser.Word(parser.Mul_rr) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func DivRR(dest, op1, op2 uint32) parser.Word {
	return parser.Word(parser.Div_rr) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func AndRI(dest, op1, op2 uint32) parser.Word {
	op2 = op2 & 0xFFFF
	return parser.Word(parser.And_ri) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func OrrRI(dest, op1, op2 uint32) parser.Word {
	op2 = op2 & 0xFFFF
	return parser.Word(parser.Orr_ri) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func XorRI(dest, op1, op2 uint32) parser.Word {
	op2 = op2 & 0xFFFF
	return parser.Word(parser.Xor_ri) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func LslRI(dest, op1, op2 uint32) parser.Word {
	op2 = op2 & 0xFFFF
	return parser.Word(parser.Lsl_ri) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func LsrRI(dest, op1, op2 uint32) parser.Word {
	op2 = op2 & 0xFFFF
	return parser.Word(parser.Lsr_ri) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func AsrRI(dest, op1, op2 uint32) parser.Word {
	op2 = op2 & 0xFFFF
	return parser.Word(parser.Asr_ri) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func MulRI(dest, op1, op2 uint32) parser.Word {
	return parser.Word(parser.Mul_ri) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func DivRI(dest, op1, op2 uint32) parser.Word {
	return parser.Word(parser.Div_ri) | parser.Word(dest)<<6 | parser.Word(op1)<<11 | parser.Word(op2)<<16
}

func Wfi() parser.Word {
	return parser.Word(parser.Wfi)
}
